url29ecd5d728_geodb_db3dcore_tgz-pJ8-de_uos_igf_db3d_dbms_model3d_SegmentNetBuilderJ8/src/de/uos/igf/db3d/dbms/util/SegmentServices.java

```diff
@@ -0,0 +1,259 @@
+ * Creates some HashMaps for a better Handling of Points from a SegmentNet: The
+ * Map "points" is a Map of unique Points with unique IDs. The Map "segments" is
+ * a Map of SegmentIDs with an Integer Array of 2 PointIDs. The Map "pointIDs"
+ * is a Map to access the Point IDs via Point objects. We need it to fill the
+ * Integer Array of Segments.
+ * 
+ * You can use this methods for Import/Export functions to sort out duplicate
+ * points.
+ * 
+ * @author Paul Vincent Kuper (kuper@kit.edu)
+ */
+public class SegmentServices {
+
+	// ID + Point3D
+	// The Map "points" is a Map of unique Points with unique IDs.
+	HashMap<Integer, Point3D> points;
+
+	// Point3D + ID
+	// The Map "pointIDs" is a Map to access the Point IDs via Point objects. We
+	// need it to fill the Integer Array of Segment.
+	HashMap<Point3D, Integer> pointIDs;
+
+	// Tetrahedron ID + 3 Point3D IDs
+	// The Map "segments" is a Map of SegmentIDs with an Integer Array
+	// of 2 PointIDs.
+	HashMap<Integer, int[]> segments;
+
+	// This Map contains all componentIDs accessible by their segmentID
+	// (segmentID, componentID)
+	HashMap<Integer, Integer> components;
+
+	/**
+	 * Constructor creates Maps and calls init function.
+	 * 
+	 */
+	public SegmentServices() {
+		points = new HashMap<Integer, Point3D>();
+		pointIDs = new HashMap<Point3D, Integer>();
+		segments = new HashMap<Integer, int[]>();
+		components = new HashMap<Integer, Integer>();
+	}
+
+	/**
+	 * The initial method to create the Maps for our services.
+	 * 
+	 * @param tetNet
+	 *            - A Segment Net
+	 */
+	public void initForPointClouds(SegmentNet3D segNet) {
+
+		SegmentNet3DComp[] comp = segNet.getComponents();
+		SegmentElt3D segment;
+		Point3D[] p;
+		int id = 0;
+
+		// save the IDs of Points to create the Segments:
+		Set<Point3D> unique = new HashSet<Point3D>();
+
+		for (SegmentNet3DComp segComp : comp) {
+
+			Set s = segComp.getElementsViaSAM();
+			Iterator<Equivalentable> it = s.iterator();
+
+			while (it.hasNext()) {
+
+				segment = (SegmentElt3D) it.next();
+
+				p = segment.getPoints();
+
+				int[] pointsForSegments = new int[2];
+
+				for (int i = 0; i < 2; i++) {
+
+					if (!unique.contains(p[i])) {
+						unique.add(p[i]);
+						points.put(id, p[i]);
+						pointIDs.put(p[i], id);
+						pointsForSegments[i] = id;
+						id++;
+					} else {
+						pointsForSegments[i] = pointIDs.get(p[i]);
+					}
+				}
+				segments.put(segment.getID(), pointsForSegments);
+				components.put(segment.getID(), segComp.getID());
+			}
+		}
+	}
+
+	/**
+	 * The initial method to create the Maps for our services.
+	 */
+	public void initFor4DPointClouds(Map<Integer, Element4D> segment4dNet,
+			Map<Integer, Point3D> pointTubes4D) {
+
+		// save the IDs of Points to create the Triangles:
+		Set<Point3D> unique = new HashSet<Point3D>();
+		
+		Point3D zero, one;		
+
+		for (Integer segmentID : segment4dNet.keySet()) {
+
+			Segment4D tmpSegment = (Segment4D) segment4dNet.get(segmentID);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.